src/librustc_ast_passes/ast_validation.rs

```diff
@@ -647,24 +647,61 @@ impl<'a> AstValidator<'a> {
             return;
         }
         // Find all generic argument coming after the first constraint...
-        let mut misplaced_args = Vec::new();
-        let mut first = None;
-        for arg in &data.args {
-            match (arg, first) {
-                (AngleBracketedArg::Arg(a), Some(_)) => misplaced_args.push(a.span()),
-                (AngleBracketedArg::Constraint(c), None) => first = Some(c.span),
-                (AngleBracketedArg::Arg(_), None) | (AngleBracketedArg::Constraint(_), Some(_)) => {
-                }
-            }
-        }
+        let constraint_spans = data
+            .args
+            .iter()
+            .filter_map(|arg| match arg {
+                AngleBracketedArg::Constraint(c) => Some(c.span),
+                _ => None,
+            })
+            .collect::<Vec<_>>();
+        let arg_spans = data
+            .args
+            .iter()
+            .filter_map(|arg| match arg {
+                AngleBracketedArg::Arg(a) => Some(a.span()),
+                _ => None,
+            })
+            .collect::<Vec<_>>();
+        let snippet_span = match &constraint_spans[..] {
+            [single] => *single,
+            [first, .., last] => first.to(*last),
+            [] => unreachable!(),
+        };
+        let removal_span = match &arg_spans[..] {
+            [first, ..] => snippet_span.until(*first),
+            [] => unreachable!(),
+        };
+        let sugg_span = match &arg_spans[..] {
+            [.., last] => last.shrink_to_hi(),
+            [] => unreachable!(),
+        };
+        let snippet = self.session.source_map().span_to_snippet(snippet_span).unwrap();
```

Comment:
Could we use `pprust` instead (clone the AST, partition it, ...)? The `.unwrap()` here doesn't seem correct and could ICE in some situations. For simplicity, (this diagnostic isn't all that commonly hit), I would prefer that we replace the whole span of the arguments (including `<` and `>`) so that we have to do less span hackery.

Reply:
That's what I ended up doing this morning before I saw this comment :)